From 97e033b3bb0cb16b61bf49f0dc7f311a3e0edd1b Mon Sep 17 00:00:00 2001
From: Ted Reed <reed@fb.com>
Date: Mon, 4 Nov 2019 07:02:54 -0800
Subject: [PATCH] Attempt to make CarbonProtocolReader::skip tail recursive

Reviewed By: edenzik

Differential Revision: D17967570

fbshipit-source-id: fdc32e190a521349c7c8f4d6081902fa18eb0284
---
 mcrouter/lib/carbon/CarbonProtocolReader.cpp | 18 +++++++++---------
 mcrouter/lib/carbon/CarbonProtocolReader.h   |  6 ++++--
 2 files changed, 13 insertions(+), 11 deletions(-)

diff --git a/mcrouter/lib/carbon/CarbonProtocolReader.cpp b/mcrouter/lib/carbon/CarbonProtocolReader.cpp
index a4cc007c..67f89c9c 100644
--- a/mcrouter/lib/carbon/CarbonProtocolReader.cpp
+++ b/mcrouter/lib/carbon/CarbonProtocolReader.cpp
@@ -78,13 +78,11 @@ void CarbonProtocolReader::skip(const FieldType ft) {
     }
     case FieldType::Struct: {
       readStructBegin();
-      while (true) {
-        const auto fieldType = readFieldHeader().first;
-        if (fieldType == FieldType::Stop) {
-          break;
-        }
-        skip(fieldType);
-      }
+      const auto next = readFieldHeader().first;
+      skip(next);
+      break;
+    }
+    case FieldType::Stop: {
       readStructEnd();
       break;
     }
@@ -96,8 +94,10 @@ void CarbonProtocolReader::skip(const FieldType ft) {
       skipKVContainer();
       break;
     }
-    default: { break; }
+    default: {
+      break;
+    }
   }
 }
 
-} // carbon
+} // namespace carbon
diff --git a/mcrouter/lib/carbon/CarbonProtocolReader.h b/mcrouter/lib/carbon/CarbonProtocolReader.h
index 5ec74365..083283f1 100644
--- a/mcrouter/lib/carbon/CarbonProtocolReader.h
+++ b/mcrouter/lib/carbon/CarbonProtocolReader.h
@@ -245,8 +245,10 @@ class CarbonProtocolReader {
   }
 
   void readStructEnd() {
-    lastFieldId_ = nestedStructFieldIds_.back();
-    nestedStructFieldIds_.pop_back();
+    if (!nestedStructFieldIds_.empty()) {
+      lastFieldId_ = nestedStructFieldIds_.back();
+      nestedStructFieldIds_.pop_back();
+    }
   }
 
   std::pair<std::pair<FieldType, FieldType>, uint32_t>
